File: 模拟进程队列/Queue.py
class Node:
    def __init__(self, data='', pre=None, post=None):
        self.data = data
        self.pre = pre
        self.post = post


class DoubleLinkedList:
    def __init__(self):
        self.head = Node()
        self.tail = Node()
        self.head.post = self.tail
        self.tail.pre = self.head
        self.length = 0

    def is_empty(self):
        return self.length == 0

    def is_in_queue(self, value=''):
        if self.is_empty():
            return False
        p = self.head
        while p.data != value and p != self.tail:
            p = p.post
        if p == self.tail:
            return False
        return True

    def append_node(self, value=''):
        if self.is_in_queue(value):
            print("进程 %s 已存在" % value)
        else:
            new_node = Node(value)
            p = self.head
            while p.post != self.tail:
                p = p.post
            new_node.post = self.tail
            new_node.pre = p
            p.post = new_node
            self.tail.pre = new_node
            self.length += 1
            print("进程 %s 入列成功" % value)

    def delete_node(self, value=''):
        if self.is_empty():
            print("进程队列为空")
            return False
        slow = self.head
        fast = self.head.post
        while fast.data != value and fast != self.tail:
            slow = slow.post
            fast = fast.post
        if fast.data == value:
            slow.post = fast.post
            fast.post.pre = slow
            self.length -= 1
            print("进程 %s 已删除" % value)
        else:
            print("进程 %s 不存在" % value)

    def in_queue(self):
        origin_str = input(">>>")
        origin_str = origin_str.split()
        for i in origin_str:
            self.append_node(i)

    def out_queue(self):
        origin_str = input(">>>")
        origin_str = origin_str.split()
        for i in origin_str:
            self.delete_node(i)

    def print_queue(self):
        p = self.head
        while p != self.tail:
            print("%s" % p.data, end=' ')
            p = p.post
        print("\n")


def show_menu():
    queue = DoubleLinkedList()
    while True:
        print('''请输入你的选择：
        \b\b1. 输入入队队列
        \b\b2. 输入出队队列
        \b\b3. 打印当前队列
        \b\b4. 离开程序''')
        choice = input('>>>')
        if choice == '1':
            print("请输入入队队列")
            queue.in_queue()
            print("\n")
        elif choice == '2':
            print("请输入出队队列")
            queue.out_queue()
            print("\n")
        elif choice == '3':
            queue.print_queue()
        elif choice == '4':
            exit(0)
        else:
            print("输入错误，请重新输入")


show_menu()

File: 模拟进程队列/Queue.c
#include <stdio.h>
#include <stdlib.h>
#define maxsize 31
#define newnode (node *)malloc(sizeof(node))


typedef struct linknode{
    char data;
    struct linknode *pre;
    struct linknode *post;
} node;


void init_node(node **new_node);
void insert_node(node *head, node *new_node);
void free_node(node **useless_node);
void free_all(node **head);
int get_queue_size(node *head);
void in_queue(node **l);
void out_queue(node **l);
void delete_queue(node *head, char ch);
void print_queue(node *head);
void init_array(char a[]);
int get_length(char a[]);
void clear_space(char a[]);
void clear_repat(char a[]);


int main(){
    node *l;
    while(1){
        char choice;
        printf("Please input your choice:\n\
            1. Input in_queue.\n\
            2. Input out_queue.\n\
            3. Print queue.\n\
            4. Exit program.\n>>>");
        fflush(stdin);
        choice = getchar();
        switch (choice){
        case '1':
            in_queue(&l);
            break;
        case '2':
            out_queue(&l);
            break;
        case '3':
            print_queue(l);
            break;
        case '4':
            free_all(&l);
            return 0;
            break;
        default:
            printf("Input ERROR!\n");
            break;
        }
    }
}


void in_queue(node **l){
    char ch[maxsize];
    init_array(ch);
    printf("Please input in_queue, you can insert %d chars.\n>>>",  maxsize - get_length(ch) -1);
    fflush(stdin);
    gets(ch);
    clear_space(ch);
    clear_repat(ch);
    int num = get_length(ch);
    init_node(l);
    for(int i=0; i<num; i++){
        node *news;
        init_node(&news);
        news->data = ch[i];
        insert_node(*l,news);
        printf("The %c in queue success !\n", ch[i]);
    }
}


void out_queue(node **l){
    char ch[maxsize];
    init_array(ch);
    int queue_size = get_queue_size(*l);
    printf("Please input out_queue, you can delete %d chars.\n>>>", queue_size);
    fflush(stdin);
    gets(ch);
    clear_space(ch);
    clear_repat(ch);
    int num = get_length(ch);
    for(int i=0; i<num; i++)
        delete_queue(*l, ch[i]);
}


void print_queue(node *head){
    int flag = 1;
    node *p = head->post;
    printf("\n");
    while(p!=NULL){
        printf("%c", p->data);
        p = p->post;
        flag = 0;
    }
    if(flag)
        printf("None");
    printf("\n\n");
}


void init_array(char a[]){
    for(int i=0; i<maxsize; i++)
        a[i] = '\0';
}


int get_length(char a[]){
    int length;
    for(length=0; a[length]!='\0'; length++);
    return length;
}


void clear_space(char a[]){
    int i, j;
    for(i=0, j=0; a[i]!='\0'; i++, j++){
        if(a[j]==32)
            j++;
        a[i] = a[j];
    }
    a[i] = '\0';
}


void clear_repat(char a[]){
    int i, j, k;
    for(i=0, k=0; a[i]!='\0'; i++, k++){
        for(j=0; j<i; j++){
            if(a[i] == a[j]){
                i++;
            }
        }
        a[k] = a[i];
    }
    a[k] = '\0';
}


void init_node(node **new_node){
    (*new_node) = newnode;
    (*new_node)->data = '\0';
    (*new_node)->pre = NULL;
    (*new_node)->post = NULL;
}


void free_node(node **useless_node){
    (*useless_node)->data='\0';
    (*useless_node)->pre = NULL;
    (*useless_node)->post = NULL;
    free(*useless_node);
}


void insert_node(node *head, node *new_node){
    while(head->post!=NULL)
        head = head->post;
    head->post = new_node;
    new_node->pre = head;
}


void delete_queue(node *head, char ch){
    node *slow = head;
    node *fast = head->post;
    while(fast!=NULL && fast->data!=ch){
        slow = slow->post;
        fast = fast->post;
    }
    if(fast == NULL)
        printf("The %c is not exist.\n", ch);
    else{
        node *temp = fast;
        slow->post = fast->post;
        fast->pre = slow;
        fast = fast->post;
        free_node(&temp);
        printf("The %c has been deleted.\n", ch);
    }
}


int get_queue_size(node *head){
    int size = 0;
    node *p = head->post;
    while(p!=NULL){
        p = p->post;
        size++;
    }
    return size;
}


void free_all(node **head){
    node *temp;
    node *p = *head;
    while(p!=NULL){
        temp = p;
        p = p->post;
        free_node(&temp);
    }
}


